### ISSUE TITLE
JSON stats blow up during timeout

### ISSUE THREAD

**Issue by bookshelfdave** [basho/riak_repl#367]
```
2013-08-09 05:01:26 =ERROR REPORT====
webmachine error: path="/stats"
{error,{exit,{json_encode,{bad_term,{error,timeout}}},[{mochijson2,json_encode,2,[{file,"src/mochijson2.erl"},{line,149}]},{mochijson2,'-json_encode_proplist/2-fun-0-',3,[{file,"src/mochijson2.erl"},{line,167}]},{lists,foldl,3,[{file,"lists.erl"},{line,1197}]},{mochijson2,json_encode_proplist,2,[{file,"src/mochijson2.erl"},{line,170}]},{riak_kv_wm_stats,produce_body,2,[{file,"src/riak_kv_wm_stats.erl"},{line,77}]},{webmachine_resource,resource_call,3,[{file,"src/webmachine_resource.erl"},{line,183}]},{webmachine_resource,do,3,[{file,"src/webmachine_resource.erl"},{line,141}]},{webmachine_decision_core,resource_call,1,[{file,"src/webmachine_decision_core.erl"},{line,48}]}]}}
```

also

```
2013-08-09 09:15:06 =ERROR REPORT====
** Generic server riak_snmp_stat_poller terminating 
** Last message in was timeout
** When Server state == {state,[{trap_state,node_get_fsm_time_mean,nodeGetTimeMean,{threshold_trap,nodeGetTimeMeanThreshold,nodeGetTimeMeanAlarmRising,nodeGetTimeMeanAlarmFalling,0,false}},{trap_state,node_get_fsm_time_median,nodeGetTimeMedian,{threshold_trap,nodeGetTimeMedianThreshold,nodeGetTimeMedianAlarmRising,nodeGetTimeMedianAlarmFalling,0,false}},{trap_state,node_get_fsm_time_95,nodeGetTime95,{threshold_trap,nodeGetTime95Threshold,nodeGetTime95AlarmRising,nodeGetTime95AlarmFalling,0,false}},{trap_state,node_get_fsm_time_99,nodeGetTime99,{threshold_trap,nodeGetTime99Threshold,nodeGetTime99AlarmRising,nodeGetTime99AlarmFalling,0,false}},{trap_state,node_get_fsm_time_100,nodeGetTime100,{threshold_trap,nodeGetTime100Threshold,nodeGetTime100AlarmRising,nodeGetTime100Alarmfalling,0,false}},{trap_state,node_put_fsm_time_mean,nodePutTimeMean,{threshold_trap,nodePutTimeMeanThreshold,nodePutTimeMeanAlarmRising,nodePutTimeMeanAlarmFalling,0,false}},{trap_state,node_put_fsm_time_median,nodePutTimeMedian,{threshold_trap,nodePutTimeMedianThreshold,nodePutTimeMedianAlarmRising,nodePutTimeMedianAlarmFalling,0,false}},{trap_state,node_put_fsm_time_95,nodePutTime95,{threshold_trap,nodePutTime95Threshold,nodePutTime95AlarmRising,nodePutTime95AlarmFalling,0,false}},{trap_state,node_put_fsm_time_99,nodePutTime99,{threshold_trap,nodePutTime99Threshold,nodePutTime99AlarmRising,nodePutTime99AlarmFalling,0,false}},{trap_state,node_put_fsm_time_100,nodePutTime100,{threshold_trap,nodePutTime100Threshold,nodePutTime100AlarmRising,nodePutTime100AlarmFalling,0,false}},{trap_state,client_connect_errors,replClientConnectErrors,{counter_trap,0,replClientConnectErrorAlarm}},{trap_state,server_connect_errors,replServerConnectErrors,{counter_trap,0,replServerConnectErrorAlarm}},{trap_state,objects_dropped_no_clients,replObjectsDroppedNoClients,{counter_trap,0,replObjectsDroppedNoClientsAlarm}},{trap_state,objects_dropped_no_leader,replObjectsDroppedNoLeader,{counter_trap,0,replObjectsDroppedNoLeaderAlarm}},{trap_state,rt_sink_errors,replRtSinkErrors,{counter_trap,0,replRtSinkErrorAlarm}},{trap_state,rt_source_errors,replRtSourceErrors,{counter_trap,0,replRtSourceErrorAlarm}}]}
** Reason for termination == 
** {timeout,{gen_server,call,[riak_repl_leader_gs,leader_node]}}
2013-08-09 09:15:06 =CRASH REPORT====
  crasher:
    initial call: riak_snmp_stat_poller:init/1
    pid: <0.3077.0>
    registered_name: riak_snmp_stat_poller
    exception exit: {{timeout,{gen_server,call,[riak_repl_leader_gs,leader_node]}},[{gen_server,terminate,6,[{file,"gen_server.erl"},{line,747}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,227}]}]}
    ancestors: [riak_snmp_sup,<0.3058.0>]
    messages: [{#Ref<0.0.469.181735>,undefined}]
    links: [<0.3076.0>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 6765
    stack_size: 24
    reductions: 3263236
  neighbours:
2013-08-09 09:15:06 =SUPERVISOR REPORT====
    Supervisor: {local,riak_snmp_sup}
    Context:    child_terminated
    Reason:    {timeout,{gen_server,call,[riak_repl_leader_gs,leader_node]}}
    Offender:  [{pid,<0.3077.0>},{name,riak_snmp_stat_poller},{mfargs,{riak_snmp_stat_poller,start_link,[]}},{restart_type,permanent},{shutdown,5000},{child_type,worker}]
```

**Comment by lordnull** [comment]
This should already be fixed in 1.4. The error is a gen_server timeout when getting the leader node for old repl (repl mode 1.2). That timeout was raised to be infinite in 82e53676776d4887ffe5176a4629baef0b08a58e. This means the snmp poller and the wm stats just block until the leader responds. Verified this behavor on develop:
1. Start a cluster
2. Add a listener (old style replication)
3. Attach to leader node
4. suspend the 'riak_repl_leader_gs' process
5. Attempt to get stats (see how it stalls)
6. resume 'riak_repl_leader_gs' prose
7. stats return.

So this opens some questions. Is this the behavior we want? If not, how do we change it? If so, do we backport to 1.3?

**Comment by bowrocker** [comment]
Closing this issue following the 2 related PRs:

https://github.com/basho/riak_repl/pull/412

and backport to 1.4.3:

https://github.com/basho/riak_repl/pull/423
